### PR TITLE
Fixing incorrect selection of ListViewGroup for ListView.GroupCollapsedStateChanged event (#6708)

### LINKED ISSUE #6708: ListView.GroupCollapsedStateChanged fires with incorrect GroupIndex/doesn't fire at all
.NET Core Version:  

* .NET Core SDK: 7.0.0-preview.2.22116.9

Have you experienced this same bug with .NET Framework?: 
No

**Problem description:**

When a key is pressed, `ListView` determines which group is affected by grabbing the `Group` property of the first selected item:

https://github.com/dotnet/winforms/blob/150aabc4cf47c303f980fbc6313c329e6da8583c/src/System.Windows.Forms/src/System/Windows/Forms/ListView.cs#L6934-L6937

This is incorrect if an item is selected in one group, followed by subsequent selection of a second group which can be done by holding down Ctrl and selecting a different group's header.

Behavior manifests different depending on whether changes in PR #6493 have been applied (note: PR not merged at time of issue creation).

- On most recent build: `ListView.GroupCollapsedStateChanged` event fires with incorrect `GroupIndex`

ListViewGroup-unpatched

- After #6493 , the event does not fire at all

ListViewGroup

**Expected behavior:**

`ListView.GroupCollapsedStateChanged` should fire and have the correct value for `GroupIndex`

**Minimal repro:**

Code used for animations:

``` C#
using var form = new Form() { Size = new Size(808, 469) };
using var listView = new ListView();
using var textBox = new TextBox() { Multiline = true };

form.Controls.Add(listView);
listView.Location = new Point(10, 10);
listView.Size = new Size(770, 200);

form.Controls.Add(textBox);
textBox.Location = new Point(10, 220);
textBox.Size = new Size(770, 200);

var group1 = new ListViewGroup("Group 1") { CollapsedState = ListViewGroupCollapsedState.Expanded };
var group2 = new ListViewGroup("Group 2") { CollapsedState = ListViewGroupCollapsedState.Expanded };

listView.Groups.AddRange(new[] { group1, group2 });
listView.Items.Add(new ListViewItem("g1 - i1", group1));
listView.Items.Add(new ListViewItem("g1 - i2", group1));
listView.Items.Add(new ListViewItem("g1 - i3", group1));
listView.Items.Add(new ListViewItem("g2 - i1", group2));
listView.Items.Add(new ListViewItem("g2 - i2", group2));
listView.Items.Add(new ListViewItem("g2 - i3", group2));

listView.GroupCollapsedStateChanged += (_, e) =>
{
	textBox.Text = $"{DateTime.Now.TimeOfDay}: CollapsedStateChanged for {listView.Groups[e.GroupIndex].Header}{Environment.NewLine}{textBox.Text}";
};

form.ShowDialog();
```

### PR DESCRIPTION
**Pull request by NikitaSemenovAkvelon**
Fixes #6708

## Proposed changes

- Processed all unique groups of selected items through collapse. The collapse state had been validated in cycle and correct group(s) triggered for GroupCollapsedStateChanged.

## Customer Impact

- ListView.GroupCollapsedStateChanged event works correctly in case of multiple groups selection.

## Regression? 

- No

## Risk

- Minimal

### Before

154516575-62709402-0a92-4ace-9579-1ea66eeef564

### After

dECVUwE2AK

## Test methodology 

- Manually tested the change in a winforms app

## Test environment(s) 

- 7.0.100-preview.5.22307.18
- Windows 11

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/7412)

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/winforms?pullRequest=7412) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/winforms?pullRequest=7412) before we can accept your contribution.<br/><br/>:x: NikitaSemenovAkvelon [sign now](https://cla.dotnetfoundation.org/dotnet/winforms?pullRequest=7412)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/winforms?pullRequest=7412) it.</sub>

**Comment by Tanya-Solyanik** [comment]
@NikitaSemenovAkvelon , @vladimir-krestov - had you guys confirmed that this bug is a regression from the .NET framework?

File: src/System.Windows.Forms/src/System/Windows/Forms/ListView.cs [hunk 1]
```diff
@@ -6929,18 +6929,22 @@ protected override void WndProc(ref Message m)
                     // User can collapse/expand a group using the keyboard by focusing the group header and using left/right
                     if (GroupsDisplayed && (key is User32.VK.LEFT or User32.VK.RIGHT) && SelectedItems.Count > 0)
                     {
-                        ListViewGroup group = SelectedItems[0].Group;
-
-                        if (group is null || group.CollapsedState is ListViewGroupCollapsedState.Default)
+                        // User can select more than one group
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/ListView.cs:6932]: nit: comments should resemble sentences, i.e. end with a period. This is not true across the code base, but we are moving in that direction.

**Review by Tanya-Solyanik — COMMENTED** [review]
Looks good, but please confirm that this is a regression from the Framework.
Is it possible to add a unit test?

**Comment by RussKie** [comment]
Please do a clean rebase on top of the latest main
image

**Comment by dreddy-work** [comment]
@NikitaSemenovAkvelon , are you planning on adding unit tests here?

**Comment by NikitaSemenovAkvelon** [comment]
> @NikitaSemenovAkvelon , are you planning on adding unit tests here?

Yes, I'm trying to add it now.

File: src/System.Windows.Forms/src/System/Windows/Forms/ListView.cs [hunk 2]
```diff
@@ -6926,21 +6926,25 @@ protected override void WndProc(ref Message m)
...
-
-                        if (group is null || group.CollapsedState is ListViewGroupCollapsedState.Default)
+                        // User can select more than one group.
+                        HashSet<int> groups = new();
+                        foreach (ListViewItem selectedItem in SelectedItems)
                         {
-                            break;
-                        }
+                            ListViewGroup group = selectedItem.Group;
+                            if (group is null || !groups.Add(group.ID) || group.CollapsedState is ListViewGroupCollapsedState.Default)
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/ListView.cs:6937]: I'd keep the original order and append the new operation to the end as it has perf implications
```suggestion
                            if (group is null || group.CollapsedState is ListViewGroupCollapsedState.Default || !groups.Add(group.ID))
```

**Review by RussKie — CHANGES_REQUESTED** [review]
👍

**Comment by RussKie** [comment]
Looking at the "after" animation - when we select both groups and collapse them together, shouldn't there be two notification, i.e., for each group?

**Comment by NikitaSemenovAkvelon** [comment]
> Looking at the "after" animation - when we select both groups and collapse them together, shouldn't there be two notification, i.e., for each group?

In animation it looks like all items are selected but the only one group.

**Comment by RussKie** [comment]
Ah, the translucent rectangle is on a specific group and it gets collapsed/expanded. 
👍

**Review by RussKie — APPROVED** [review]
LGTM

**Review by Tanya-Solyanik — APPROVED** [review]

**Comment by Tanya-Solyanik** [comment]
@vladimir-krestov - I assume you are adding test in a separate PR?

**Comment by vladimir-krestov** [comment]
> @vladimir-krestov - I assume you are adding test in a separate PR?

Yes, will try

**Comment by vladimir-krestov** [comment]
> @vladimir-krestov - I assume you are adding test in a separate PR?

PR #7489

**Pull request merged**